unittests/block_state_tests.cpp

```diff
@@ -0,0 +1,111 @@
+                     (id)
+                     (header)
+                     (pending_schedule)
+                     (activated_protocol_features)
+                     (additional_signatures)
+)
+
+
+BOOST_AUTO_TEST_SUITE(block_state_tests)
+
+BOOST_AUTO_TEST_CASE(test_unpack_legacy_block_state) {
+   eosio::testing::tester main;
+
+   using namespace eosio::chain::literals;
+
+   // First we create a valid block with valid transaction
+   main.create_account("newacc"_n);
+   auto b = main.produce_block();
+
+   auto bs = main.control->head_block_state();
+
+   // pack block_header_state as the legacy format
+   fc::datastream<std::vector<char>> out_strm;
+   fc::raw::pack(out_strm, reinterpret_cast<eosio::chain::block_header_state_v0&>(*bs));
+
+   BOOST_CHECK_NE(out_strm.storage().size(), 0);
+
+   // make sure we can unpack block_header_state 
+   {
+      fc::datastream<const char*> in_strm(out_strm.storage().data(), out_strm.storage().size());
+      eosio::chain::block_header_state_unpack_stream unpack_strm(in_strm, 0);
+      eosio::chain::block_header_state tmp;
+      BOOST_CHECK_NO_THROW(fc::raw::unpack(unpack_strm, tmp));
+      BOOST_CHECK_EQUAL(bs->id, tmp.id);
+      BOOST_CHECK_EQUAL(in_strm.remaining(), 0);
+   }
+
+   // manual pack legacy block_state
+   fc::raw::pack(out_strm, bs->block);
+   fc::raw::pack(out_strm, false);
+
+   // make sure we can unpack block_state 
+   {
+      fc::datastream<const char*> in_strm(out_strm.storage().data(), out_strm.storage().size());
+      eosio::chain::block_header_state_unpack_stream unpack_strm(in_strm, 0);
+      eosio::chain::block_state tmp;
+      BOOST_CHECK_NO_THROW(fc::raw::unpack(unpack_strm, tmp));
+      BOOST_CHECK_EQUAL(bs->id, tmp.id);
+      BOOST_CHECK_EQUAL(in_strm.remaining(), 0);
+   }
+}
+
+BOOST_AUTO_TEST_CASE(test_unpack_new_block_state) {
+eosio::testing::tester main;
+
+   using namespace eosio::chain::literals;
+
+   // First we create a valid block with valid transaction
+   main.create_account("newacc"_n);
+   auto b = main.produce_block();
+
+   auto bs = main.control->head_block_state();
+   auto& sgi = bs->get_security_group_info();
+   sgi.version = 1;
+   sgi.participants.insert("adam"_n);
+
+   // pack block_header_state as the legacy format
+   fc::datastream<std::vector<char>> out_strm;
+   fc::raw::pack(out_strm, *bs);
+
+   BOOST_CHECK_NE(out_strm.storage().size(), 0);
+
+
+   // make sure we can unpack block_state 
+   {
+      fc::datastream<const char*> in_strm(out_strm.storage().data(), out_strm.storage().size());
+      eosio::chain::block_header_state_unpack_stream unpack_strm(in_strm, 1);
+      eosio::chain::block_state tmp;
+      BOOST_CHECK_NO_THROW(fc::raw::unpack(unpack_strm, tmp));
+      BOOST_CHECK_EQUAL(bs->id, tmp.id);
+      BOOST_CHECK_EQUAL(bs->get_security_group_info().version, 1);
+      BOOST_TEST(bs->get_security_group_info().participants == sgi.participants);
+      BOOST_CHECK_EQUAL(in_strm.remaining(), 0);
```

Comment:
Should also check something on the bs->block to validate that the block_state data was correctly unpacked.